ably/state.go

```diff
@@ -0,0 +1,374 @@
+package ably
+
+import (
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/ably/ably-go/ably/proto"
+)
+
+// StateType specifies which group of states is relevant in given context;
+// either:
+//
+//   - StateConn* group describing Conn states
+//   - StateChan* group describing RealtimeChannel states
```

Comment:
You will need to have then two separate State’s, ChanState and ConnState and in order to do that you’ll need to duplicate code for stateEmitter to handle two distinct types. Having this as just int not only allows you to write the code only once, it enables the user to blanket handle all state transitions with single function (see ClientOptions.Listener).